File: programs/zmart-core/src/instructions/initialize_global_config.rs
use anchor_lang::prelude::*;
use crate::state::GlobalConfig;
use crate::error::ErrorCode;

/// Initialize the global protocol configuration (one-time operation)
///
/// This instruction creates the GlobalConfig PDA with default settings for:
/// - Fee structure (3% protocol, 2% resolver, 5% LP)
/// - Voting thresholds (70% proposal approval, 60% dispute success)
/// - Time limits (24h minimum resolution delay, 3 days dispute period)
/// - Protocol admin and wallets
///
/// Can only be called once. Subsequent calls will fail with AlreadyInitialized.
///
/// # Arguments
///
/// * `backend_authority` - Backend service authority for vote aggregation
///
/// # Errors
///
/// * `InvalidFeeConfiguration` - If fee sum exceeds 10000 basis points (100%)
#[derive(Accounts)]
pub struct InitializeGlobalConfig<'info> {
    /// Protocol administrator who pays for account creation and has full control
    #[account(mut)]
    pub admin: Signer<'info>,

    /// Global configuration PDA (created once)
    ///
    /// Seeds: [b"global-config"]
    /// Space: GlobalConfig::LEN (198 bytes + 8 discriminator)
    #[account(
        init,
        seeds = [b"global-config"],
        bump,
        payer = admin,
        space = 8 + GlobalConfig::LEN
    )]
    pub global_config: Account<'info, GlobalConfig>,

    /// Wallet receiving protocol fees (3% of trading fees)
    /// CHECK: Any account can be protocol wallet, validated by admin
    pub protocol_fee_wallet: UncheckedAccount<'info>,

    /// System program for account creation
    pub system_program: Program<'info, System>,
}

/// Handler for initialize_global_config instruction
pub fn handler(
    ctx: Context<InitializeGlobalConfig>,
    backend_authority: Pubkey,
) -> Result<()> {
    let global_config = &mut ctx.accounts.global_config;

    // Set admin and backend authority
    global_config.admin = ctx.accounts.admin.key();
    global_config.backend_authority = backend_authority;

    // Set protocol fee wallet
    global_config.protocol_fee_wallet = ctx.accounts.protocol_fee_wallet.key();

    // Set fee configuration (3% + 2% + 5% = 10% total trading fee)
    global_config.protocol_fee_bps = 300; // 3%
    global_config.resolver_reward_bps = 200; // 2%
    global_config.liquidity_provider_fee_bps = 500; // 5%

    // Validate fee configuration
    let total_fees = global_config.protocol_fee_bps
        .checked_add(global_config.resolver_reward_bps)
        .ok_or(ErrorCode::OverflowError)?
        .checked_add(global_config.liquidity_provider_fee_bps)
        .ok_or(ErrorCode::OverflowError)?;

    require!(
        total_fees <= 10000, // Max 100%
        ErrorCode::InvalidFeeConfiguration
    );

    // Set voting thresholds
    global_config.proposal_approval_threshold = 7000; // 70%
    global_config.dispute_success_threshold = 6000; // 60%

    // Set time limits (in seconds)
    global_config.min_resolution_delay = 86_400; // 24 hours
    global_config.dispute_period = 259_200; // 3 days (72 hours)

    // Set minimum resolver reputation (80%)
    global_config.min_resolver_reputation = 8000;

    // Initialize state
    global_config.is_paused = false;
    global_config.bump = ctx.bumps.global_config;

    // Initialize reserved space to zero
    global_config.reserved = [0; 64];

    msg!(
        "Global config initialized by admin: {}",
        ctx.accounts.admin.key()
    );

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn create_test_global_config() -> GlobalConfig {
        GlobalConfig {
            admin: Pubkey::new_unique(),
            backend_authority: Pubkey::new_unique(),
            protocol_fee_wallet: Pubkey::new_unique(),
            protocol_fee_bps: 300,
            resolver_reward_bps: 200,
            liquidity_provider_fee_bps: 500,
            proposal_approval_threshold: 7000,
            dispute_success_threshold: 6000,
            min_resolution_delay: 86_400,
            dispute_period: 259_200,
            min_resolver_reputation: 8000,
            is_paused: false,
            reserved: [0; 64],
            bump: 255,
        }
    }

    #[test]
    fn test_default_fee_configuration() {
        let config = create_test_global_config();

        // Verify 3/2/5 split (3% protocol, 2% resolver, 5% LP)
        assert_eq!(config.protocol_fee_bps, 300);
        assert_eq!(config.resolver_reward_bps, 200);
        assert_eq!(config.liquidity_provider_fee_bps, 500);

        // Total should be 10% (1000 bps)
        let total = config.protocol_fee_bps + config.resolver_reward_bps + config.liquidity_provider_fee_bps;
        assert_eq!(total, 1000);
    }

    #[test]
    fn test_default_voting_thresholds() {
        let config = create_test_global_config();

        // Proposal approval requires 70%
        assert_eq!(config.proposal_approval_threshold, 7000);

        // Dispute success requires 60%
        assert_eq!(config.dispute_success_threshold, 6000);
    }

    #[test]
    fn test_default_time_limits() {
        let config = create_test_global_config();

        // Minimum resolution delay: 24 hours
        assert_eq!(config.min_resolution_delay, 86_400);
        assert_eq!(config.min_resolution_delay, 24 * 60 * 60);

        // Dispute period: 3 days
        assert_eq!(config.dispute_period, 259_200);
        assert_eq!(config.dispute_period, 3 * 24 * 60 * 60);
    }

    #[test]
    fn test_fee_configuration_validation() {
        // Valid: 3% + 2% + 5% = 10% < 100%
        let valid_total = 300u16 + 200u16 + 500u16;
        assert_eq!(valid_total, 1000);
        assert!(valid_total <= 10000);

        // Invalid: would exceed 100%
        let invalid_protocol = 9000u16;
        let invalid_resolver = 2000u16;
        let invalid_total = invalid_protocol + invalid_resolver;
        assert!(invalid_total > 10000);
    }

    #[test]
    fn test_initial_state() {
        let config = create_test_global_config();

        // Protocol should not be paused initially
        assert_eq!(config.is_paused, false);

        // Min resolver reputation should be 80%
        assert_eq!(config.min_resolver_reputation, 8000);

        // Reserved space should be zeroed
        assert_eq!(config.reserved, [0; 64]);
    }
}

File: tests/devnet/initialize-global-config.ts
import * as anchor from "@coral-xyz/anchor";
import { Program } from "@coral-xyz/anchor";
import { ZmartCore } from "../target/types/zmart_core";
import { SystemProgram, PublicKey, LAMPORTS_PER_SOL } from "@solana/web3.js";
import { assert } from "chai";

describe("Initialize Global Config", () => {
  // Configure the client to use devnet
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);

  // Load program manually instead of using workspace
  const programId = new PublicKey("7h3gXfBfYFueFVLYyfL5Qo1QGsf4GQUfW96FKVgnUsJS");
  const idl = require("../target/idl/zmart_core.json");
  const program = new Program(idl, programId, provider) as Program<ZmartCore>;

  const admin = provider.wallet.publicKey;

  console.log("\n" + "=".repeat(60));
  console.log("Global Config Initialization Test");
  console.log("=".repeat(60));
  console.log(`Program ID: ${program.programId.toString()}`);
  console.log(`Admin: ${admin.toString()}`);
  console.log(`Cluster: ${provider.connection.rpcEndpoint}`);

  // Derive global config PDA
  const [globalConfigPDA, bump] = PublicKey.findProgramAddressSync(
    [Buffer.from("global-config")],
    program.programId
  );

  console.log(`Global Config PDA: ${globalConfigPDA.toString()}`);
  console.log(`Bump: ${bump}`);

  it("Checks wallet balance", async () => {
    const balance = await provider.connection.getBalance(admin);
    console.log(`\nWallet Balance: ${(balance / LAMPORTS_PER_SOL).toFixed(4)} SOL`);

    assert.isTrue(
      balance > 0.1 * LAMPORTS_PER_SOL,
      "Insufficient balance! Need at least 0.1 SOL for initialization"
    );
  });

  it("Checks if global config already exists", async () => {
    try {
      const existingConfig = await provider.connection.getAccountInfo(globalConfigPDA);

      if (existingConfig) {
        console.log("\n⚠️  Global Config already exists!");
        console.log(`   Size: ${existingConfig.data.length} bytes`);
        console.log(`   Owner: ${existingConfig.owner.toString()}`);

        // Try to fetch and display the config
        try {
          const config = await program.account.globalConfig.fetch(globalConfigPDA);
          console.log("\n📋 Existing Config:");
          console.log(`   Admin: ${config.admin.toString()}`);
          console.log(`   Backend Authority: ${config.backendAuthority.toString()}`);
          console.log(`   Protocol Fee Wallet: ${config.protocolFeeWallet.toString()}`);
          console.log(`   Protocol Fee: ${config.protocolFeeBps} bps (${config.protocolFeeBps / 100}%)`);
          console.log(`   Resolver Fee: ${config.resolverRewardBps} bps (${config.resolverRewardBps / 100}%)`);
          console.log(`   LP Fee: ${config.liquidityProviderFeeBps} bps (${config.liquidityProviderFeeBps / 100}%)`);
          console.log(`   Paused: ${config.isPaused}`);

          console.log("\n✅ Skipping initialization (already complete)");
          this.skip();
        } catch (fetchError) {
          console.error("   Error fetching config data:", fetchError);
          throw new Error("Global config exists but cannot be decoded!");
        }
      } else {
        console.log("\n✅ Global Config does not exist yet - ready to initialize");
      }
    } catch (error) {
      // RPC errors are expected when account doesn't exist
      if (error.message?.includes("could not find account")) {
        console.log("\n✅ Global Config does not exist yet - ready to initialize");
      } else {
        throw error;
      }
    }
  });

  it("Initializes global config with default parameters", async () => {
    console.log("\n" + "=".repeat(60));
    console.log("Initializing Global Config...");
    console.log("=".repeat(60));

    const backendAuthority = admin; // Can be different in production
    const protocolFeeWallet = admin; // Can be different in production

    console.log(`\nParameters:`);
    console.log(`   Admin: ${admin.toString()}`);
    console.log(`   Backend Authority: ${backendAuthority.toString()}`);
    console.log(`   Protocol Fee Wallet: ${protocolFeeWallet.toString()}`);

    console.log(`\nDefault Configuration:`);
    console.log(`   Protocol Fee: 3% (300 bps)`);
    console.log(`   Resolver Fee: 2% (200 bps)`);
    console.log(`   LP Fee: 5% (500 bps)`);
    console.log(`   Total Trading Fee: 10% (1000 bps)`);
    console.log(`   Proposal Approval Threshold: 70% (7000 bps)`);
    console.log(`   Dispute Success Threshold: 60% (6000 bps)`);
    console.log(`   Min Resolution Delay: 24 hours (86400s)`);
    console.log(`   Dispute Period: 3 days (259200s)`);
    console.log(`   Min Resolver Reputation: 80% (8000 bps)`);

    try {
      console.log(`\n⏳ Sending transaction...`);

      const tx = await program.methods
        .initializeGlobalConfig(backendAuthority)
        .accounts({
          admin: admin,
          globalConfig: globalConfigPDA,
          protocolFeeWallet: protocolFeeWallet,
          systemProgram: SystemProgram.programId,
        })
        .rpc();

      console.log(`\n✅ Transaction confirmed!`);
      console.log(`   Signature: ${tx}`);
      console.log(`   Explorer: https://explorer.solana.com/tx/${tx}?cluster=devnet`);

      // Wait for confirmation
      console.log(`\n⏳ Waiting for account data to be available...`);
      await new Promise(resolve => setTimeout(resolve, 3000));

    } catch (error) {
      console.error(`\n❌ Transaction failed:`, error);

      if (error.logs) {
        console.log(`\n📋 Program Logs:`);
        error.logs.forEach((log, i) => console.log(`   ${i + 1}. ${log}`));
      }

      // Check if it's an "already initialized" error
      if (error.message?.includes("already in use") || error.message?.includes("custom program error: 0x0")) {
        console.log(`\n⚠️  Account already initialized - this is OK!`);
        console.log(`   Continuing to verification...`);
      } else {
        throw error;
      }
    }
  });

  it("Verifies global config was initialized correctly", async () => {
    console.log("\n" + "=".repeat(60));
    console.log("Verifying Initialization...");
    console.log("=".repeat(60));

    // Fetch the global config account
    const config = await program.account.globalConfig.fetch(globalConfigPDA);

    console.log(`\n✅ Global Config Account Fetched Successfully!`);
    console.log(`\n📋 Configuration Details:`);
    console.log(`   PDA: ${globalConfigPDA.toString()}`);
    console.log(`   Bump: ${config.bump}`);
    console.log(`\n👤 Authorities:`);
    console.log(`   Admin: ${config.admin.toString()}`);
    console.log(`   Backend: ${config.backendAuthority.toString()}`);
    console.log(`   Protocol Fee Wallet: ${config.protocolFeeWallet.toString()}`);
    console.log(`\n💰 Fee Structure:`);
    console.log(`   Protocol: ${config.protocolFeeBps} bps (${config.protocolFeeBps / 100}%)`);
    console.log(`   Resolver: ${config.resolverRewardBps} bps (${config.resolverRewardBps / 100}%)`);
    console.log(`   LP: ${config.liquidityProviderFeeBps} bps (${config.liquidityProviderFeeBps / 100}%)`);

    const totalFee = config.protocolFeeBps + config.resolverRewardBps + config.liquidityProviderFeeBps;
    console.log(`   Total: ${totalFee} bps (${totalFee / 100}%)`);

    console.log(`\n🗳️  Voting Thresholds:`);
    console.log(`   Proposal Approval: ${config.proposalApprovalThreshold} bps (${config.proposalApprovalThreshold / 100}%)`);
    console.log(`   Dispute Success: ${config.disputeSuccessThreshold} bps (${config.disputeSuccessThreshold / 100}%)`);

    console.log(`\n⏱️  Time Limits:`);
    console.log(`   Min Resolution Delay: ${config.minResolutionDelay}s (${config.minResolutionDelay / 3600}h)`);
    console.log(`   Dispute Period: ${config.disputePeriod}s (${config.disputePeriod / 86400}d)`);

    console.log(`\n⭐ Reputation:`);
    console.log(`   Min Resolver: ${config.minResolverReputation} bps (${config.minResolverReputation / 100}%)`);

    console.log(`\n🔐 State:`);
    console.log(`   Paused: ${config.isPaused}`);

    // Assertions
    assert.equal(config.admin.toString(), admin.toString(), "Admin mismatch");
    assert.equal(config.protocolFeeBps, 300, "Protocol fee should be 3%");
    assert.equal(config.resolverRewardBps, 200, "Resolver fee should be 2%");
    assert.equal(config.liquidityProviderFeeBps, 500, "LP fee should be 5%");
    assert.equal(totalFee, 1000, "Total fee should be 10%");
    assert.equal(config.proposalApprovalThreshold, 7000, "Proposal threshold should be 70%");
    assert.equal(config.disputeSuccessThreshold, 6000, "Dispute threshold should be 60%");
    assert.equal(config.minResolutionDelay.toNumber(), 86400, "Min resolution delay should be 24h");
    assert.equal(config.disputePeriod.toNumber(), 259200, "Dispute period should be 3 days");
    assert.equal(config.minResolverReputation, 8000, "Min resolver reputation should be 80%");
    assert.equal(config.isPaused, false, "Should not be paused initially");

    console.log(`\n✅ All validations passed!`);
  });

  it("Verifies account can be fetched via RPC", async () => {
    console.log(`\n⏳ Fetching account via RPC...`);

    const accountInfo = await provider.connection.getAccountInfo(globalConfigPDA);

    assert.isNotNull(accountInfo, "Global config account should exist");
    assert.equal(accountInfo.owner.toString(), program.programId.toString(), "Wrong owner");
    assert.isTrue(accountInfo.data.length > 0, "Account should have data");

    console.log(`\n✅ RPC Verification Successful!`);
    console.log(`   Owner: ${accountInfo.owner.toString()}`);
    console.log(`   Size: ${accountInfo.data.length} bytes`);
    console.log(`   Lamports: ${(accountInfo.lamports / LAMPORTS_PER_SOL).toFixed(6)} SOL`);
  });

  after(() => {
    console.log("\n" + "=".repeat(60));
    console.log("🎉 Global Config Initialization Complete!");
    console.log("=".repeat(60));
    console.log(`\n📋 Summary:`);
    console.log(`   PDA: ${globalConfigPDA.toString()}`);
    console.log(`   Admin: ${admin.toString()}`);
    console.log(`   Ready for: Market creation and trading`);
    console.log(`\n💡 Next Steps:`);
    console.log(`   1. Create a test market`);
    console.log(`   2. Test trading on devnet`);
    console.log(`   3. Test frontend integration`);
    console.log("=".repeat(60) + "\n");
  });
});
